### PR TITLE
qa: Add basic mypy support for the qa directory

### PR DESCRIPTION
**Pull request by toabctl**
This is useful to catch basic errors.

### CODE REVIEW

File: qa/mypy.ini [hunk 1]
```diff
@@ -0,0 +1,2 @@
+[mypy]
```
Comment (sebastian-philipp) [inline qa/mypy.ini]: do we really need so many mypy.ini files? I though one file (`ceph.git/src/mypy.ini`) would be enough for everyone.

File: qa/tasks/ceph_manager.py [hunk 2]
```diff
@@ -996,7 +996,7 @@ def wrapper(self):
                 raise
         return wrapper
 
-    @log_exc
+    @log_exc  # type: ignore
```
Comment (sebastian-philipp) [inline qa/tasks/ceph_manager.py]: which error do you get here?

File: qa/tasks/cephfs/test_full.py [hunk 3]
```diff
@@ -375,8 +375,8 @@ class TestQuotaFull(FullnessTestCase):
     """
     Test per-pool fullness, which indicates quota limits exceeded
     """
-    pool_capacity = 1024 * 1024 * 32   # arbitrary low-ish limit
-    fill_mb = pool_capacity / (1024 * 1024)
+    pool_capacity = 1024 * 1024 * 32  # type: ignore # arbitrary low-ish limit
```
Comment (sebastian-philipp) [inline qa/tasks/cephfs/test_full.py]: again, which error do you get here?

File: qa/tasks/ceph_manager.py (hunk 2 above)
Comment (toabctl) [reply qa/tasks/ceph_manager.py]: `error: Argument 1 to "log_exc" has incompatible type "Callable[[OSDThrasher], Any]"; expected "OSDThrasher"`

File: qa/tasks/cephfs/test_full.py (hunk 3 above)
Comment (toabctl) [reply qa/tasks/cephfs/test_full.py]: `error: Incompatible types in assignment (expression has type "int", base class "FullnessTestCase" defined the type as "None")`

File: qa/mypy.ini (hunk 1 above)
Comment (toabctl) [reply qa/mypy.ini]: With that one, I get `Found 2442 errors in 99 files (checked 194 source files)`. So this needs to be incrementally fixed imo. That's why I added a new file.

File: qa/tasks/ceph_manager.py (hunk 2 above)
Comment (sebastian-philipp) [reply qa/tasks/ceph_manager.py]: Yeah, turns out, `log_exc` is a non-static method with the self-argument renamed to `func` and we're using the unbound `log_exc` method and passing the decorated function as the self-argument. 

So, we could either make `log_exc` a `staticmethod` and then use something like 

```python
class C:
     @staticmethod
     def deco(func):
         def w(self):
             func(self)
         return w

     @deco.__get__(object)
     def func(self):
         print(2)
```

or we can move log_exc to top-level. (my preferred solution)

File: qa/tasks/cephfs/test_full.py (hunk 3 above)
Comment (sebastian-philipp) [reply qa/tasks/cephfs/test_full.py]: proper fix would be do change FullnessTestCase:

```python
class FullnessTestCase:
    pool_capacity = None  # type: Optional[int]
```

File: qa/tasks/ceph_manager.py (hunk 2 above)
Comment (toabctl) [reply qa/tasks/ceph_manager.py]: Done

File: qa/tasks/cephfs/test_full.py (hunk 3 above)
Comment (toabctl) [reply qa/tasks/cephfs/test_full.py]: Done. Thanks!

File: qa/tasks/ceph_manager.py [hunk 4]
```diff
@@ -110,11 +110,21 @@ def mount_osd_data(ctx, remote, cluster, osd):
...
+            self.log(traceback.format_exc())
+            raise
+    return wrapper
+
+
 class PoolType:
     REPLICATED = 1
     ERASURE_CODED = 3
 
-
```
Comment (sebastian-philipp) [inline qa/tasks/ceph_manager.py]: As this is toplevel,  I'd leave this blank linke to stay pep8 compatible.

Comment (toabctl) [reply qa/tasks/ceph_manager.py]: ah. sorry. fixed

File: qa/tasks/ceph_manager.py [hunk 5]
```diff
@@ -110,6 +110,17 @@ def mount_osd_data(ctx, remote, cluster, osd):
             )
 
 
+def log_exc(func):
```
Comment (tchaikov) [inline qa/tasks/ceph_manager.py]: just out of curiosity, why do we need to move `log_exc()` to the top level of this file?

File: qa/tasks/ceph_objectstore_tool.py [hunk 6]
```diff
@@ -514,7 +514,7 @@ def test_objectstore(ctx, config, cli_remote, REP_POOL, REP_NAME, ec=False):
                 proc.wait()
                 if proc.exitstatus != 0:
                     log.error("Failure of --op info command with {ret}".
-                              format(proc.exitstatus))
+                              format(ret=proc.exitstatus))
```
Comment (tchaikov) [inline qa/tasks/ceph_objectstore_tool.py]: could you split changes like this into another commit? also, i'd suggest use the f-string, like:
```py
log.error(f"Failure of --op info command with {proc.exitstatus}")
```

File: qa/tasks/cephfs/filesystem.py [hunk 7]
```diff
@@ -583,8 +583,8 @@ def check_pool_application(self, pool_name):
             if pool['pool_name'] == pool_name:
                 if "application_metadata" in pool:
                     if not "cephfs" in pool['application_metadata']:
-                        raise RuntimeError("Pool %p does not name cephfs as application!".\
-                                           format(pool_name))
+                        raise RuntimeError("Pool {pool_name} does not name cephfs as application!".\
```
Comment (tchaikov) [inline qa/tasks/cephfs/filesystem.py]: ditto.

File: qa/tasks/cephfs/test_scrub_checks.py [hunk 8]
```diff
@@ -167,10 +167,9 @@ def _checks(self, run_seq):
         log.info("Cloning repo into place")
         repo_path = TestScrubChecks.clone_repo(self.mount_a, client_path)
 
-        log.info("Initiating mds_scrub_checks on mds.{id_}, " +
-                 "test_path {path}, run_seq {seq}".format(
-                     id_=mds_rank, path=abs_test_path, seq=run_seq)
-                 )
+        log.info("Initiating mds_scrub_checks on mds.{id_} test_path {path}, run_seq {seq}".format(
```
Comment (tchaikov) [inline qa/tasks/cephfs/test_scrub_checks.py]: i think the reason why we wrapped the line is to please requirement like https://www.python.org/dev/peps/pep-0008/#id19 .

File: qa/tasks/cephfs/test_snapshots.py [hunk 9]
```diff
@@ -189,9 +189,9 @@ def test_kill_mdstable(self):
         rank1 = self.fs.get_rank(rank=1, status=status)
         self.fs.rank_freeze(True, rank=0)
         self.fs.rank_freeze(True, rank=1)
-        self.fs.rank_asok(['config', 'set', "mds_kill_mdstable_at", "8".format(i)], rank=0, status=status)
-        self.fs.rank_asok(['config', 'set', "mds_kill_mdstable_at", "3".format(i)], rank=1, status=status)
-        proc = self.mount_a.run_shell(["mkdir", "d1/dir/.snap/s4".format(i)], wait=False)
+        self.fs.rank_asok(['config', 'set', "mds_kill_mdstable_at", "8"], rank=0, status=status)
```
Comment (tchaikov) [inline qa/tasks/cephfs/test_snapshots.py]: @batrick could you help review this part?

File: qa/tasks/ceph_manager.py (hunk 5 above)
Comment (sebastian-philipp) [reply qa/tasks/ceph_manager.py]: the previous method misused the self argument as the function. Like so:

```python
In [3]: class C:
     def decorator(self):
         def wrapper(arg):
             return self(arg)
         return wrapper
     @decorator
     def func(self): ...
```

unfortunately that actually works in Python.

Comment (tchaikov) [reply qa/tasks/ceph_manager.py]: > the previous method misused the self argument as the function. Like so:

i don't agree. "self" is but a parameter name for the wrapped function from the perspective of `log_exc()`. and `log_exc()` assumes that the wrapped function only accepts a single parameter. it could be anything, but here, it is always "self". that's why i used "self" as the name of the parameter.

i don't see anything wrong or inappropriate here.

> Edit: mypy expects the `self` argument to have the same type as the class like so:

"self" does have the same type of the class.

File: qa/tasks/ceph_objectstore_tool.py (hunk 6 above)
Comment (toabctl) [reply qa/tasks/ceph_objectstore_tool.py]: Why splitting? It is fixing problems detected by mypy.

File: qa/tasks/cephfs/test_scrub_checks.py (hunk 8 above)
Comment (toabctl) [reply qa/tasks/cephfs/test_scrub_checks.py]: But we dont' force it (eg. with flake8) so I don't see a problem here. Or do you want me to change it back?

File: qa/tasks/ceph_manager.py (hunk 5 above)
Comment (sebastian-philipp) [reply qa/tasks/ceph_manager.py]: > > the previous method misused the self argument as the function. Like so:
> 
> i don't agree. "self" is but a parameter name for the wrapped function from the perspective of `log_exc()`. and `log_exc()` assumes that the wrapped function only accepts a single parameter. it could be anything, but here, it is always "self". that's why i used "self" as the name of the parameter.

I disagree here. Common sense is that the first argument of class methods are called `self`, because self is handled differently than other parameters. As `log_exc` expected a different type passed in as the first argument, it should live outside of the class. IMO.

> 
> i don't see anything wrong or inappropriate here.
> 
> > Edit: mypy expects the `self` argument to have the same type as the class like so:
> 
> "self" does have the same type of the class.

exactly. That's why it should be moved to toplevel.

File: qa/tasks/cephfs/test_scrub_checks.py (hunk 8 above)
Comment (tchaikov) [reply qa/tasks/cephfs/test_scrub_checks.py]: i don't feel strong on this. just wanted to make sure we are improving the quality of code or or making it compliant to a sound policy/guideline.

File: qa/tasks/ceph_objectstore_tool.py (hunk 6 above)
Comment (toabctl) [reply qa/tasks/ceph_objectstore_tool.py]: fstrings are python3 only but we currently still have python2

**Comment by toabctl** [comment]
as requested by @tchaikov , PR is now splitted into 2 commits. please have another look.

File: qa/tasks/cephfs/test_snapshots.py (hunk 9 above)
Comment (batrick) [reply qa/tasks/cephfs/test_snapshots.py]: Looks good, was just a harmless copy paste bug apparently.

**Comment by tchaikov** [comment]
jenkins test make check

File: qa/tasks/ceph_manager.py [hunk 10]
```diff
@@ -986,6 +986,7 @@ def do_thrash(self):
             # Allow successful completion so gevent doesn't see an exception.
             # The DaemonWatchdog will observe the error and tear down the test.
 
+    @staticmethod
```
Comment (tchaikov) [inline qa/tasks/ceph_manager.py]: we need to update the caller site. otherwise these "log_exc" will be unbound method. in other words, should put.
```py
    @OSDThrasher.log_exc
    def do_sighup(self):
      # ...
```

```
TypeError: 'staticmethod' object is not callable
2020-01-13T14:45:58.296 DEBUG:teuthology.run_tasks:Unwinding manager install
2020-01-13T14:45:58.311 ERROR:teuthology.contextutil:Saw exception from nested tasks
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/contextutil.py", line 34, in nested
    yield vars
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/task/install/__init__.py", line 599, in task
    yield
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 86, in run_tasks
    manager = run_one_task(taskname, ctx=ctx, config=config)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 64, in run_one_task
    task = get_task(taskname)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 28, in get_task
    module = _import('tasks', module_name, task_name, fail_on_import_error=True)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 52, in _import
    0,
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/gevent/builtins.py", line 96, in __import__
    result = _import(*args, **kwargs)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2020-01-13-1112/qa/tasks/ceph.py", line 21, in <module>
    from ceph_manager import CephManager, write_conf
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/gevent/builtins.py", line 96, in __import__
    result = _import(*args, **kwargs)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2020-01-13-1112/qa/tasks/ceph_manager.py", line 118, in <module>
    class OSDThrasher(Thrasher):
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2020-01-13-1112/qa/tasks/ceph_manager.py", line 1004, in OSDThrasher
    @log_exc
TypeError: 'staticmethod' object is not callable
```

see http://pulpito.ceph.com/kchai-2020-01-13_06:08:15-rados-wip-kefu-testing-2020-01-13-1112-distro-basic-smithi/4664265/

Comment (toabctl) [reply qa/tasks/ceph_manager.py]: @tchaikov done. please retest.

File: qa/tasks/ceph_manager.py [hunk 11]
```diff
@@ -1055,7 +1056,7 @@ def wrapper(self):
                 raise
         return wrapper
 
-    @log_exc
+    @OSDThrasher.log_exc
```
Comment (tchaikov) [inline qa/tasks/ceph_manager.py]:
```
2020-01-24T10:31:28.462 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 86, in run_tasks
    manager = run_one_task(taskname, ctx=ctx, config=config)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 64, in run_one_task
    task = get_task(taskname)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 28, in get_task
    module = _import('tasks', module_name, task_name, fail_on_import_error=True)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 52, in _import
    0,
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/gevent/builtins.py", line 96, in __import__
    result = _import(*args, **kwargs)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2020-01-23-2212/qa/tasks/ceph.py", line 21, in <module>
    from ceph_manager import CephManager, write_conf
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/gevent/builtins.py", line 96, in __import__
    result = _import(*args, **kwargs)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2020-01-23-2212/qa/tasks/ceph_manager.py", line 123, in <module>
    class OSDThrasher(Thrasher):
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2020-01-23-2212/qa/tasks/ceph_manager.py", line 1066, in OSDThrasher
    @OSDThrasher.log_exc
NameError: name 'OSDThrasher' is not defined
```

http://pulpito.ceph.com/kchai-2020-01-24_02:23:44-rados-wip-kefu-testing-2020-01-23-2212-distro-basic-smithi/4698061/

Comment (toabctl) [reply qa/tasks/ceph_manager.py]: I moved the `log_exc` method to top level again. @tchaikov Could you retest please?

**Comment by toabctl** [comment]
@sebastian-philipp could you have a look again?

File: qa/.gitignore [hunk 12]
```diff
@@ -2,3 +2,4 @@
 .*.sw[nmop]
 *.pyc
 .tox
+.mypy_cache
```
Comment (sebastian-philipp) [inline qa/.gitignore]: remove this hunk. See #33510

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2020-02-29_17:36:34-rados-wip-kefu-testing-2020-02-29-2106-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-03-01_09:33:43-rados-wip-kefu-testing-2020-03-01-1346-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-03-02_05:23:09-rados-wip-kefu-testing-2020-03-01-2348-distro-basic-smithi/

File: qa/.gitignore (hunk 12 above)
Comment (tchaikov) [reply qa/.gitignore]: @toabctl could you address this comment?

Comment (toabctl) [reply qa/.gitignore]: Done

**Comment by toabctl** [comment]
@tchaikov @sebastian-philipp rebased and comment addressed.
